### PR TITLE
ref: Set sample rates to default if out of range

### PR DESCRIPTION
**Pull request by philipphofmann**
## :scroll: Description

Set sampleRate, tracesSampleRate, and tracesSampler to default value when
a provided value is out of range.

## :bulb: Motivation and Context

Fixes GH-986

## :green_heart: How did you test it?
Unit tests.

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the CHANGELOG if needed
- [x] I updated the docs if needed
- [x] Review from the native team if needed
- [ ] No breaking changes

### CODE REVIEW

File: Sources/Sentry/SentryOptions.m [hunk 1]
```diff
@@ -25,14 +33,16 @@ - (instancetype)init
...
+        _defaultSampleRate = @1;
+        self.sampleRate = _defaultSampleRate;
         self.enableAutoSessionTracking = YES;
         self.enableOutOfMemoryTracking = YES;
         self.sessionTrackingIntervalMillis = [@30000 unsignedIntValue];
         self.attachStacktrace = YES;
         self.maxAttachmentSize = 20 * 1024 * 1024;
         self.sendDefaultPii = NO;
-        self.tracesSampleRate = @0;
+        _defaultTracesSampleRate = @0;
```
Comment (bruno-garcia) [inline Sources/Sentry/SentryOptions.m:44]: I believe the default here should be `nil`.

`0` means: Don't randomly create new transactions but if a trace has already been sampled, then honor that sampling decision.

File: Sources/Sentry/SentryOptions.m [hunk 2]
```diff
@@ -25,14 +33,16 @@ - (instancetype)init
         self.maxBreadcrumbs = defaultMaxBreadcrumbs;
         self.maxCacheItems = 30;
         self.integrations = SentryOptions.defaultIntegrations;
-        self.sampleRate = @1;
+        _defaultSampleRate = @1;
```
Comment (bruno-garcia) [inline Sources/Sentry/SentryOptions.m:36]: Shouldn't this just be a const defined inline somewhere?

File: Sources/Sentry/SentryOptions.m [hunk 3]
```diff
@@ -230,4 +239,26 @@ - (void)addInAppExclude:(NSString *)inAppExclude
...
+    } else {
+        _sampleRate = _defaultSampleRate;
+    }
+}
+
+- (void)setTracesSampleRate:(NSNumber *)tracesSampleRate
+{
+    if (tracesSampleRate == nil) {
+        _tracesSampleRate = nil;
+    } else if ([tracesSampleRate doubleValue] >= 0 && [tracesSampleRate doubleValue] <= 1.0) {
```
Comment (bruno-garcia) [inline Sources/Sentry/SentryOptions.m:257]: L: What about we assign `[tracesSampleRate doubleValue]` and just use that value  on the second hand of `&&`?

File: Sources/Sentry/TracesSampler.m [hunk 4]
```diff
@@ -23,13 +23,17 @@ - (instancetype)initWithOptions:(SentryOptions *)options
...
+    if (context.transactionContext.sampled != kSentrySampleDecisionUndecided) {
         return context.transactionContext.sampled;
+    }
 
     if (_options.tracesSampler != nil) {
         NSNumber *callbackDecision = _options.tracesSampler(context);
         if (callbackDecision != nil)
-            return [self calcSample:callbackDecision.doubleValue];
+            if ([callbackDecision doubleValue] < 0 || [callbackDecision doubleValue] > 1.0) {
+                callbackDecision = @0;
```
Comment (bruno-garcia) [inline Sources/Sentry/TracesSampler.m:34]: What about `nil`?

`0` means: Honor trace sampling decision

File: Sources/Sentry/SentryOptions.m (hunk 2 above)
Comment (philipphofmann) [reply Sources/Sentry/SentryOptions.m]: Inline not, I also need it for the setters. Unfortunately, you can't use NSNumber for constants. That's a drawback of Objective-C. We could do some weird stuff as described here https://stackoverflow.com/questions/12268904/nsnumber-constants-in-obj-c, but instead, I prefer to use private properties instead.

File: Sources/Sentry/SentryOptions.m (hunk 3 above)
Comment (philipphofmann) [reply Sources/Sentry/SentryOptions.m]: Done

**Review by bruno-garcia — APPROVED** [review]

**Pull request merged**